ethcontract/src/transaction/build.rs

```diff
@@ -230,6 +220,31 @@ async fn build_offline_signed_transaction<T: Transport>(
     Ok(signed.raw_transaction)
 }
 
+async fn resolve_gas_limit<T: Transport>(
+    web3: &Web3<T>,
+    from: Address,
+    gas_price: GasPrice,
+    options: &TransactionOptions,
+) -> Result<U256, ExecutionError> {
+    match options.gas {
+        Some(value) => Ok(value),
+        None => Ok(web3
+            .eth()
+            .estimate_gas(
+                CallRequest {
+                    from: Some(from),
+                    to: options.to,
+                    gas: None,
```

Comment:
I think specifying `None` as a `gas_limit` in the `estimate_gas` call itself can also cause issues. I noticed that https://rpc.xdaichain.com/ returned a huge limit (>13M) for simple ~200k gas txs which meant the tx wasn't picked up by any miner. However with https://dai.poa.network as a node, things worked fine.

To work around that, we could specify the block's gas limit (can be queried from current block) as a value here. In that case https://rpc.xdaichain.com/ returned a better, yet still way too high estimate (2M instead of 200k). However, this would cost another roundtrip to the node and would have to be updated in the `method.estimate_gas call` as well.

Reply:
Hmm, sounds tricky...

> returned a huge limit (>13M)

Sounds like it decided to return block gas limit? Weird...

Reply:
The [block gas limit is 12.5M on xDAI](https://blockscout.com/poa/xdai/blocks/14843345/transactions) so that estimate leads to signing a tx that cannot be mined. [Dani mentioned] in slack it could be related to using nethermind as a Node.